sample_keyprovider/src/enc_mods/cc_kbc_enc/mod.rs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use aes_gcm::aead::{Aead, KeyInit};
+use aes_gcm::{Aes256Gcm, Key, Nonce};
+use anyhow::*;
+use rand::RngCore;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+pub struct AnnotationPacket {
+    // Key Resource ID (URL)
+    // Format:
+    // `cc_kbc://127.0.0.1:8080/test_repo/key/id_1`
+    pub kid: String,
+    // Encrypted key to unwrap (base64-encoded)
+    pub wrapped_data: String,
+    // Initialisation vector (base64-encoded)
+    pub iv: String,
+    // Wrap type to specify encryption algorithm and mode
+    pub wrap_type: String,
+}
+
+pub fn aes_gcm_256_encrypt(data: &[u8], key: &[u8], iv: &[u8]) -> Result<Vec<u8>> {
+    let encryption_key = Key::<Aes256Gcm>::from_slice(key);
+    let cipher = Aes256Gcm::new(encryption_key);
+    let nonce = Nonce::from_slice(iv);
+    let encrypted_data = cipher
+        .encrypt(nonce, data.as_ref())
+        .map_err(|e| anyhow!("Decrypt failed: {:?}", e))?;
+
+    Ok(encrypted_data)
+}
+
+pub fn enc_optsdata_gen_anno(optsdata: &[u8], params: Vec<String>) -> Result<String> {
+    let (key_file_path, kid) = params[0].split_once(':').ok_or(anyhow!(
+        "Failed to parse parameters: {:?}, need key file path and key URL in KBS split by ':'",
+        params
+    ))?;
+
+    let key = std::fs::read(key_file_path).map_err(|e| anyhow!("Read Key file failed: {}", e))?;
+
+    let mut iv: [u8; 12] = [0; 12];
```

Comment:
I think that the number `12` can be replaced by `Aes256Gcm::NonceSize` (haven't tested).

Reply:
I found that `aes-gcm` crate does not provide such a constant. The `NonceSize` provided by it is an `ArrayLength<u8>` trait: https://docs.rs/aes-gcm/0.10.1/aes_gcm/type.Nonce.html?search=NonceSize. Therefore, it cannot be used as use here. Perhaps it is more appropriate to use a constant defined by ourselves like:
```
const AES_256_GCM_NONCE_SIZE: usize = 12;
```

Reply:
Yes, you're right. The constant you added is good.

----

My side research:

I found that if you add:
```
use aes_gcm::AeadCore;
```
then you get a generate_nonce function:
```
let nonce = Aes256Gcm::generate_nonce(OsRng);
```
which must return the correct length for Aes256Gcm. Then we can just pass this Aes256Gcm::Nonce variable to the encrypt function (instead of `iv: &[u8]`) -- but I'm actually having trouble finding the correct type for such an argument.

Reply:
That's great. The return type of `Aes256Gcm::generate_nonce` is `Aes256Gcm::Nonce`, which is actually `GenericArray<u8, NonceSize>`. We can use `to_vec()` to convert this struct to `Vec<u8>`.